test/captcha.test.js

```diff
@@ -6,87 +6,157 @@ const lockOpts = {
+      beforeEach(function(done) {
+        this.stub = h.stubGetChallenge([svgCaptchaRequiredResponse1, svgCaptchaRequiredResponse2]);
+        this.lock = h.displayLock('', lockOpts, done);
+      });
 
-    afterEach(function() {
-      this.lock.hide();
-    });
+      afterEach(function() {
+        this.lock.hide();
+      });
 
-    it('should show the captcha input', function() {
-      expect(h.qInput(this.lock, 'captcha', false)).to.be.ok();
-    });
+      it('should show the captcha input', function() {
+        expect(h.qInput(this.lock, 'captcha', false)).to.be.ok();
+      });
 
-    it('should require another challenge when clicking the refresh button', function(done) {
-      h.clickRefreshCaptchaButton(this.lock);
-      setTimeout(() => {
-        expect(h.q(this.lock, '.auth0-lock-captcha-image').style.backgroundImage).to.equal(
-          `url("${requiredResponse2.image}")`
-        );
-        done();
-      }, 200);
-    });
+      it('should require another challenge when clicking the refresh button', function(done) {
+        h.clickRefreshCaptchaButton(this.lock);
+        setTimeout(() => {
+          expect(h.q(this.lock, '.auth0-lock-captcha-image').style.backgroundImage).to.equal(
+            `url("${svgCaptchaRequiredResponse2.image}")`
+          );
+          done();
+        }, 200);
+      });
+
+      it('should submit the captcha provided by the user', function() {
+        h.logInWithEmailPasswordAndCaptcha(this.lock);
+        expect(h.wasLoginAttemptedWith({ captcha: 'captchaValue' })).to.be.ok();
+      });
 
-    it('should submit the captcha provided by the user', function() {
-      h.logInWithEmailPasswordAndCaptcha(this.lock);
-      expect(h.wasLoginAttemptedWith({ captcha: 'captchaValue' })).to.be.ok();
+      it('should not submit the form if the captcha is not provided', function() {
+        h.logInWithEmailAndPassword(this.lock);
+        expect(h.wasLoginAttemptedWith({})).to.not.be.ok();
+      });
     });
 
-    it('should not submit the form if the captcha is not provided', function() {
-      h.logInWithEmailAndPassword(this.lock);
-      expect(h.wasLoginAttemptedWith({})).to.not.be.ok();
+    describe('when the challenge api returns required: false', function() {
+      beforeEach(function(done) {
+        h.stubGetChallenge({
+          required: false
+        });
+        this.lock = h.displayLock('', lockOpts, done);
+      });
+
+      afterEach(function() {
+        this.lock.hide();
+      });
+
+      it('should not show the captcha input', function() {
+        expect(h.qInput(this.lock, 'captcha', false)).to.not.be.ok();
+      });
+
+      describe('when the form submission fails and the transaction starts requiring a challenge', function() {
+        beforeEach(function(done) {
+          h.assertAuthorizeRedirection((lockID, options, authParams, cb) => {
+            cb(new Error('bad request'));
+            setTimeout(done, 300);
```

Comment:
this feels dirty. is there no better way to call `done`?

Reply:
This is required because we wait 250ms to handle errors. I wasn't able to use Sinon to setup a fake timer because this is an integration test.

https://github.com/auth0/lock/blob/a79d7b7061fd78314630808d6b53b7bc21840370/src/core/actions.js#L189